lib/geoip/geoip.go

```diff
@@ -0,0 +1,167 @@
+// Copyright (C) 2024 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+// Package geoip provides an automatically updating MaxMind GeoIP2 database
+// provider.
+package geoip
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"path"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/oschwald/geoip2-golang"
+)
+
+const maxDatabaseSize = 1 << 30 // 1 GiB, at the time of writing the database is about 95 MiB
+
+type Provider struct {
+	edition         string
+	licenseKey      string
+	refreshInterval time.Duration
+	directory       string
+
+	mut        sync.Mutex
+	db         *geoip2.Reader
+	lastOpened time.Time
+}
+
+// NewGeoLite2CityProvider returns a new GeoIP2 database provider for the
+// GeoLite2-City database. The database will be stored in the given
+// directory (which should exist) and refreshed every 7 days.
+func NewGeoLite2CityProvider(licenseKey string, directory string) *Provider {
+	return &Provider{
+		edition:         "GeoLite2-City",
+		licenseKey:      licenseKey,
+		refreshInterval: 7 * 24 * time.Hour,
+		directory:       directory,
+	}
+}
+
+func (p *Provider) City(ip net.IP) (*geoip2.City, error) {
```

Comment:
This might download the entire db, right? Seems a bit unexpected. What about doing it in the constructor, and then afterwards in the background to avoid taking a long time calling this.

Reply:
Indeed